### ISSUE TITLE
Types incorrectly inferred when using a lambda that captures the environment

### ISSUE THREAD

**Issue by dob9601** [microsoft/pyright#2343]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Pyright will complain about types if a variable that could hold multiple types is used within a lambda, even if a check has been done beforehand to ensure type safety

**To Reproduce**
```py
from typing import Callable, Optional

def takes_callable(func: Callable):
    return func()

def optional_returner() -> Optional[int]:
    return 5

x = optional_returner()   # x has type of int or None
if x is not None:  # x can no longer be None within this if statement
    takes_callable(lambda: x + 5)  # Operator "+" not supported for "None"
```

**Expected behavior**
Pyright not to raise an error

**VS Code extension or command-line**
coc-pyright (command-line)

**Additional context**
pyright v1.1.171

**Comment by erictraut** [comment]
Type narrowing cannot be applied to captured variables because the type of the variable could change by the time the lambda is called.

Consider the following:
```
if x is not None:
    takes_callable(lambda: x + 5)
x = None
```

A lambda is a different execution context, and a type checker cannot generally determine when it will be executed.

The recommended workaround is to assign the narrowed type to a temporary variable whose type is guaranteed not to change after it is captured.

```python
if x is not None:
    y = x
    takes_callable(lambda: y + 5)
```

**Comment by dob9601** [comment]
I see, thank you for letting me know.
I'm guessing there's no easy way for pyright to tell as to whether the lambda is only used within the if statements because of pythons scoping rules

**Comment by erictraut** [comment]
Yes, it's possible that the callee could save it and call it at some later time.

This same issue has been reported multiple times in the recent past. For example, see [here](https://github.com/microsoft/pyright/issues/2335). I agree that the current behavior is confusing. 

Thinking about this a bit more, perhaps there's a way that using the narrowed type could be provably safe in certain circumstances. In particular, it should be safe if all assignments to the expression occur in the code flow prior to the point of variable capture. Currently pyright has no notion of "point of capture"; it simply knows about independent execution contexts. In the specific case of a lambda, the point of capture should be easy to determine.

I'll do some experimentation and see if this approach is feasible.

**Comment by erictraut** [comment]
I found a way to reliably apply type narrowing to a local variable or parameter that is captured by an inner-scoped lambda or function. The narrowed type is allowed only in cases where the type analyzer can prove that there is no code path that assigns a new value to the symbol after the lambda or def statement.

This will eliminate the need to use a temporary variable as I suggested above.

This functionality will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.172, which I just published. It will also be in the next release of pylance.
